src/main/java/com/google/firebase/database/connection/WebsocketConnection.java

```diff
@@ -338,67 +312,102 @@ public void onClose() {
         logger.debug("closed");
       }
       if (!isClosed) {
-        executorService.execute(new Runnable() {
-          @Override
-          public void run() {
-            onClosed();
-          }
-        });
+        // If the connection tear down was initiated by the higher-layer, isClosed will already
+        // be true. Nothing more to do in that case.
+        executorService.execute(
+            new Runnable() {
+              @Override
+              public void run() {
+                onClosed();
+              }
+            });
       }
     }
 
     @Override
-    public void onError(final WebSocketException e) {
+    public void onError(final Throwable e) {
+      if (e.getCause() != null && e.getCause() instanceof EOFException) {
+        logger.error("WebSocket reached EOF", e);
+      } else {
+        logger.error("WebSocket error", e);
+      }
       executorService.execute(
           new Runnable() {
             @Override
             public void run() {
-              if (e.getCause() != null && e.getCause() instanceof EOFException) {
-                logger.debug("WebSocket reached EOF.");
-              } else {
-                logger.debug("WebSocket error.", e);
-              }
               onClosed();
             }
           });
     }
+  }
 
-    @Override
-    public void onLogMessage(String msg) {
-      if (logger.logsDebug()) {
-        logger.debug("Tubesock: " + msg);
-      }
-    }
+  private static class StringList {
 
-    @Override
-    public void send(String msg) {
-      ws.send(msg);
+    private final AtomicInteger remaining = new AtomicInteger(0);
+    private List<String> buffer;
+
+    void initialize(int capacity) {
+      remaining.set(capacity);
+      buffer = new ArrayList<>(capacity);
     }
 
-    @Override
-    public void close() {
-      ws.close();
+    int append(String frame) {
+      checkState(hasRemaining());
+      buffer.add(frame);
+      return remaining.decrementAndGet();
     }
 
-    private void shutdown() {
-      ws.close();
-      try {
-        ws.blockClose();
-      } catch (InterruptedException e) {
-        logger.error("Interrupted while shutting down websocket threads", e);
-      }
+    boolean hasRemaining() {
+      return remaining.get() > 0;
     }
 
-    @Override
-    public void connect() {
+    String combine() {
```

Comment:
This deservers a better name or at least a comment :) Thanks!

Reply:
Added comment.